go/lib/snet/scion_layer.go

```diff
@@ -0,0 +1,273 @@
+	if err != nil {
+		return err
+	}
+	copy(bytes, headerBytes)
+
+	l4PayloadBytes := s.L4
+	l4PayloadBytes = append(l4PayloadBytes, s.RawPayload...)
+	bytes, err = b.AppendBytes(len(l4PayloadBytes))
+	if err != nil {
+		return err
+	}
+	copy(bytes, l4PayloadBytes)
+	return nil
+}
+
+func (s *SCIONLayer) serializeExtnForAuth(b gopacket.SerializeBuffer) error {
+	for i := 0; i < len(s.E2E); i++ {
+		bytes, err := b.PrependBytes(len(s.E2E[i].AuthenticatedBytes))
+		if err != nil {
+			return err
+		}
+		copy(bytes, s.E2E[i].AuthenticatedBytes)
+	}
+	bytes, err := b.PrependBytes(len(s.SecExt.AuthenticatedBytes))
+	if err != nil {
+		return err
+	}
+	copy(bytes, s.SecExt.AuthenticatedBytes)
+	for i := 0; i < len(s.HBH); i++ {
+		bytes, err := b.PrependBytes(len(s.HBH[i].AuthenticatedBytes))
+		if err != nil {
+			return err
+		}
+		copy(bytes, s.HBH[i].AuthenticatedBytes)
+	}
+
+	return nil
+
+}
+
+func (s *SCIONLayer) serializeCommonHdr() error {
+	hdrLen := spkt.CmnHdrLen + len(s.AddrHdr) + len(s.RawPath)
+	packtLen := len(s.HBHRaw) + len(s.E2ERaw) + len(s.SecExtnBuffer) + len(s.L4) + len(s.RawPayload)
+	//still using this struct, because it's convenient and does the job
+	s.CmnHdr = spkt.CmnHdr{
+		Ver:       spkt.SCIONVersion,
+		DstType:   s.Destination.Host.Type(),
+		SrcType:   s.Source.Host.Type(),
+		TotalLen:  uint16(hdrLen + packtLen),      //buffer should contain:l4 header, payload, extensions
+		HdrLen:    uint8(hdrLen / common.LineLen), //in multiple of 8
+		CurrInfoF: 0,
+		CurrHopF:  0,
+		NextHdr:   s.nextHdr[len(s.nextHdr)-1],
+	}
+
+	s.CmnHdrRaw = make(common.RawBytes, spkt.CmnHdrLen)
+	s.CmnHdr.Write(s.CmnHdrRaw)
+	return nil
+}
+
+func (s *SCIONLayer) serializeExtensions() error {
+	StableSortExtensions(s.Extensions)
+	hbh, e2e, err := hpkt.ValidateExtensions(s.Extensions)
+	if err != nil {
+		return err
+	}
+
+	s.E2ERaw, err = s.serializeExtensionsHelper(e2e)
+	if err != nil {
+		return nil
+	}
+	if s.AuthExt != nil {
+		s.SecExt, err = layers.ExtensionDataToSPSELayer(s.nextHdr[len(s.nextHdr)-1], *s.AuthExt)
+		if err != nil {
+			return err
+		}
+		s.SecExtnBuffer = s.SecExt.Serialize()
+		//s.SecExt.SetAuthenticator(common.RawBytes{255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255})
+		s.nextHdr = append(s.nextHdr, s.AuthExt.Class())
+	}
+	s.HBHRaw, err = s.serializeExtensionsHelper(hbh)
+	if err != nil {
+		return err
+	}
+	return nil
+
+}
+
+func (s *SCIONLayer) serializeExtensionsHelper(extensions []common.Extension) (common.RawBytes, error) {
+	raw := common.RawBytes{}
+	for i := len(extensions) - 1; i >= 0; i-- {
+		extnLayer, err := layers.ExtensionDataToExtensionLayer(s.nextHdr[len(s.nextHdr)-1], extensions[i])
+		if err != nil {
+			return nil, err
+		}
+
+		bf := gopacket.NewSerializeBuffer()
+		err = extnLayer.SerializeTo(bf, gopacket.SerializeOptions{FixLengths: true})
+		if err != nil {
+			return nil, err
+		}
+		switch extensions[i].Class() {
```

Comment:
we should not need this distinction. The caller knows that `extensions` where e2e or hbh, just return buffers and remove these internal `e.HBH`, etc.